apps/emqx_mq/src/emqx_mq_state_storage.erl

```diff
@@ -0,0 +1,539 @@
+    emqx_ds_pmap:fold(
+        fun(Shard, Progress, Acc) ->
+            Acc#{Shard => Progress}
+        end,
+        #{},
+        ShardProgress
+    ).
+
+commit_consumer_state(_NeedClaimOwnership = false, #{?collection_dirty := false} = CSRec) ->
+    {ok, CSRec};
+commit_consumer_state(
+    NeedClaimOwnership,
+    CSRec0 = #{id := Id}
+) ->
+    TxRes = emqx_ds:trans(
+        trans_opts(Id),
+        fun() -> persist_consumer_state_tx(NeedClaimOwnership, CSRec0) end
+    ),
+    case TxRes of
+        {atomic, TXSerial, CSRec} when NeedClaimOwnership ->
+            {ok, CSRec#{?collection_guard := TXSerial}};
+        {atomic, _, CSRec} ->
+            {ok, CSRec};
+        {nop, CSRec} ->
+            {ok, CSRec};
+        Err ->
+            {error, {failed_to_commit_consumer_state, Err}}
+    end.
+
+-spec destroy_consumer_state(emqx_mq_types:mq_handle()) -> ok.
+destroy_consumer_state(MQHandle) ->
+    Id = consumer_state_id(MQHandle),
+    TxRes = emqx_ds:trans(
+        trans_opts(Id),
+        fun() ->
+            emqx_ds_pmap:tx_delete_guard(Id),
+            emqx_ds_pmap:tx_destroy(Id, ?pn_shard_progress)
+        end
+    ),
+    case TxRes of
+        {atomic, _TXSerial, _} ->
+            ok;
+        {nop, _} ->
+            ok;
+        Err ->
+            {error, {failed_to_destroy_consumer_state, Err}}
+    end.
+
+-spec create_mq_state(emqx_mq_types:mq()) -> {ok, mq_state()} | {error, term()}.
+create_mq_state(MQ) ->
+    Id = mq_state_id(MQ),
+    TransOpts = trans_opts(Id),
+    TxRes = emqx_ds:trans(
+        TransOpts,
+        fun() ->
+            case open_mq_state_tx(Id) of
+                undefined ->
+                    MQStateNewRec0 = put_mq(MQ, new_mq_state(MQ)),
+                    %% Token is used just as collection presence marker.
+                    %% We do not have concept of MQ state ownership.
+                    emqx_ds_pmap:tx_write_guard(Id, ?ds_tx_serial),
+                    persist_mq_state_tx(MQStateNewRec0);
+                FoundMQStateRec ->
+                    FoundMQStateRec
+            end
+        end
+    ),
+    case TxRes of
+        {atomic, TXSerial, MQStateRec} ->
+            {ok, MQStateRec#{?collection_guard := TXSerial}};
+        {nop, MQStateRec} ->
+            {ok, MQStateRec};
+        Err ->
+            {error, {failed_to_create_mq_state, Err}}
+    end.
+
+-spec update_mq_state(emqx_mq_types:mqid(), map()) ->
+    {ok, emqx_mq_types:mq()} | not_found | {error, term()}.
+update_mq_state(MQId, MQFields) ->
+    Id = mq_state_id(MQId),
+    TransOpts = trans_opts(Id),
+    TxRes = emqx_ds:trans(
+        TransOpts,
+        fun() ->
+            case open_mq_state_tx(Id) of
+                undefined ->
+                    not_found;
+                MQStateRec0 ->
+                    MQ0 = emqx_ds_pmap:collection_get(?col_mq, ?mq_key, MQStateRec0),
+                    MQ = maps:merge(MQ0, MQFields),
```

Comment:
Q: should we filter out the immutable fields (id, topic_filter, is_lastvalue, ...) here?

Reply:
This module is considered to be low-level, not for direct usage by other parts of the app; the filtering is in `emqx_mq_registry` module.